ddtrace/contrib/cherrypy/middleware.py

```
@@ -0,0 +1,114 @@
+from ...propagation.http import HTTPPropagator
+
+# 3p
+import cherrypy
+from cherrypy.lib.httputil import valid_status
+
+log = logging.getLogger(__name__)
+
+SPAN_NAME = "cherrypy.request"
+
+
+class TraceTool(cherrypy.Tool):
+    def __init__(self, app, tracer, service, use_distributed_tracing):
+        self.app = app
+        self._tracer = tracer
+        self._service = service
+        self._use_distributed_tracing = use_distributed_tracing
+
+        cherrypy.Tool.__init__(self, "on_start_resource", self._on_start_resource, priority=95)
+
+    def _setup(self):
+        cherrypy.Tool._setup(self)
+        cherrypy.request.hooks.attach("on_end_request", self._on_end_request, priority=5)
+        cherrypy.request.hooks.attach("after_error_response", self._after_error_response, priority=5)
+
+    def _on_start_resource(self):
+        if self._use_distributed_tracing:
+            propagator = HTTPPropagator()
+            context = propagator.extract(cherrypy.request.headers)
+            # Only need to active the new context if something was propagated
+            if context.trace_id:
+                self._tracer.context_provider.activate(context)
+
+        try:
+            cherrypy.request._datadog_span = self._tracer.trace(
+                SPAN_NAME,
+                service=self._service,
+                span_type=SpanTypes.WEB,
+            )
+        except Exception:
+            log.debug("cherrypy: error tracing request", exc_info=True)
+
+    def _after_error_response(self):
+        error_type = cherrypy._cperror._exc_info()[0]
+        error_msg = str(cherrypy._cperror._exc_info()[1])
+        error_tb = cherrypy._cperror.format_exc()
+
+        self.close_span(error_type, error_msg, error_tb)
+
+    def _on_end_request(self):
+        self.close_span()
+
+    def close_span(self, error_type=None, error_msg=None, error_tb=None):
+        span = getattr(cherrypy.request, "_datadog_span", None)
+
+        if not span or not span.sampled:
```

If there's no span here then it's probably an error with the instrumentation which might lead to memory leaks (if spans are started and not finished). It's probably worth adding a `log.warning` here for when that happens.

---

Yes! I've handled that now with a warning as oppose to what would have been a silent error.